### PR TITLE
Add Documentation for Testing E2E Resource Changes

### PR DESCRIPTION
**Pull request by harrryr**
*Issue #, if available:*
This commit will aim to provide developers with a guide for setting up their personal testing environment running E2E tests

*Description of changes:*
- README.md file under ./testing that describes how to setup github and aws account to run EKS and EC2 end-to-end tests
- Changed e2e-eks-test.yml to download enablement script from the public [repo ](aws-observability/application-signals-demo) instead of a S3 bucket
- Modified testing/terraform/ec2/main.tf file to make it easier to setup testing environment
- Added extra secrets for EC2 S3 bucket since bucket names are global and each developer will need to choose a unique s3 bucket on their account. 

Test run: 
- EC2: https://github.com/harrryr/aws-otel-java-instrumentation/actions/runs/7548731778
- EKS: https://github.com/harrryr/aws-otel-java-instrumentation/actions/runs/7548756009

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: testing/README.md [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+### 1. Create an IAM Role with OIDC Identity Provider
+This step is needed to allow Github Action to have access to resources in the AWS account
+#### Create an OIDC Provider
+- First step is to create an OIDC Identity Provider to allow Github action access to the AWS account resource. Login to AWS, go to the IAM console and click on the Identity Providers tab.
+- Click on Add Provider, choose OpenID Connect and type `https://token.actions.githubusercontent.com` in the Provider URL. Click "Get thumbprint". For Audience, use `sts.amazonaws.com`. Finally, click "Add provider"
+#### Create an IAM role
+- Next, an IAM role needs to be created using the OIDC Identity Provider. Go to the Roles tab and click Create role. 
+- Choose Web Identity, and choose `token.actions.githubusercontent.com` as the Identity provider, Audience as `sts.amazonaws.com`, and for Github organizations put your github username down. Click next.
+- Add the AdministratorAccess policy. Click next.
+- Enter your Role name. Click "Create role".
```
Comment (majanjua-amzn) [inline testing/README.md:13]: Can't we do this programatically?

File: testing/README.md [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+#### Setup Environment with the Appropriate Roles and Permissions.
+- First, assume Admin role from the test account by running `ada credentials update --account=<AccountID> --role=Admin --provider=isengard --once`
+- Assume the e2e test role by running 
+  - `output=$(aws sts assume-role --role-arn arn:aws:iam::<AccountID>:role/<E2ETestRole> --role-session-name AWSCLI-Session)`
+  - `export AWS_ACCESS_KEY_ID=$(echo $output | jq -r .Credentials.AccessKeyId)`
+  - `export AWS_SECRET_ACCESS_KEY=$(echo $output | jq -r .Credentials.SecretAccessKey)`
+  - `export AWS_SESSION_TOKEN=$(echo $output | jq -r .Credentials.SessionToken)`
+- Run `aws sts get-caller-identity` to check if you are in the correct role
+#### Create a new Cluster
+- Next, create the cluster by running `eksctl create cluster --name <ClusterName> --region us-east-1 --zones us-east-1a,us-east-1b`. This will take around ~10 minutes.
```
Comment (majanjua-amzn) [inline testing/README.md:39]: Always point out the reader needs to replace something in the command before the command itself. Room for error here

File: testing/README.md [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+  - `curl -O https://raw.githubusercontent.com/kubernetes-sigs/aws-load-balancer-controller/v2.4.7/docs/install/iam_policy.json`
+  - `aws iam create-policy --policy-name AWSLoadBalancerControllerIAMPolicy --policy-document file://iam_policy.json --region us-east-1`
+  - `eksctl create iamserviceaccount --cluster=<ClusterName> --namespace=kube-system --name=aws-load-balancer-controller --attach-policy-arn=arn:aws:iam::<AccountID>:policy/AWSLoadBalancerControllerIAMPolicy --region us-east-1 --approve`
+  - `kubectl apply --validate=false -f https://github.com/jetstack/cert-manager/releases/download/v1.5.4/cert-manager.yaml`
+  - `curl -Lo v2_4_7_full.yaml https://github.com/kubernetes-sigs/aws-load-balancer-controller/releases/download/v2.4.7/v2_4_7_full.yaml`
+  - `sed -i.bak -e '561,569d' ./v2_4_7_full.yaml`
+  - `sed -i.bak -e 's|your-cluster-name|<ClusterName>|' ./v2_4_7_full.yaml`
+  - `kubectl apply -f v2_4_7_full.yaml`
+  - `curl -Lo v2_4_7_ingclass.yaml https://github.com/kubernetes-sigs/aws-load-balancer-controller/releases/download/v2.4.7/v2_4_7_ingclass.yaml`
+  - `kubectl apply -f v2_4_7_ingclass.yaml`
```
Comment (majanjua-amzn) [inline testing/README.md:52]: Can we use a code block here instead of bullet points and inline code?

File: testing/README.md [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+- Type the role name and click "Create role".
+- 
+#### Setting Up Default VPC
+- Go to the VPC console and on the routing table for the default VPC, click Edit routes. 
+- Click add routes, for destination add `0.0.0.0/0`, for target add Internet Gateway and save changes.
+- Go to the Security groups tab, find the security group attached to the default VPC, click Edit inbound rules, choose type: All Traffic, Source: custom, and CIDR block: 0.0.0.0/0. Save rules.
+
+### 4. Building Sample App to ECR
+Create two ECR repositories: one for the sample app main service and another for the sample app remote service. 
+Follow the instructions under ./sample-apps/README.md to build the sample app image and upload it to the ECR
```
Comment (majanjua-amzn) [inline testing/README.md:68]: IIRC there's a way to link between pages instead of giving the path

File: testing/README.md [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+
+### 5. Building Sample App to S3 Bucket
+Create an S3 Bucket to store the .jar files for the sample app main service and sample app remote service.
+Follow the instructions under ./sample-apps/README.md to build the sample app .jar and upload it to the bucket
```
Comment (majanjua-amzn) [inline testing/README.md:72]: Same as above

File: testing/README.md [hunk 6]
```diff
@@ -0,0 +1,95 @@
+# How to Test E2E Resource Changes
+This guide will give a step by step instruction on how to test changes made to E2E testing resources before pushing a PR. 
+
```
Comment (majanjua-amzn) [inline testing/README.md:3]: Could be useful to add a message/list of what this guide will entail, or even a table of contents, e.g. "To test changes to the E2E tests, you will need to fork the repo, set up IAM roles and an EKS cluster, and build sample apps into ECRs/S3 buckets" and anything else of relevance.

File: testing/README.md [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+  - APP_SIGNALS_E2E_EC2_TEST_ROLE: <EC2_IAM_ROLE_NAME>
+  - APP_SIGNALS_E2E_FE_SA_IMG: `<AccountID>.dkr.ecr.us-east-1.amazonaws.com/<Path to Sample App Image>`
+  - APP_SIGNALS_E2E_RE_SA_IMG: `<AccountID>.dkr.ecr.us-east-1.amazonaws.com/<Path to Remote Sample App Image>`
+  - APP_SIGNALS_E2E_FE_SA_JAR: s3://<BucketName>/<FileName.jar>
+  - APP_SIGNALS_E2E_RE_SA_JAR: s3://<BucketName>/<FileName.jar>
+
+
+### 7. Running the tests
+Copy paste the test.yml into `../.github/workflows` and replace the cluster name with the one generated in step 2. 
+Push the code changes and there should be a test running on the forked repo in the Action tab
```
Comment (majanjua-amzn) [inline testing/README.md:89]: The tester should just be able to run the canary on their test environment. Is that doable without having to make edits?

File: testing/test.yml
Comment (majanjua-amzn) [inline testing/test.yml:1]: I don't think this file is necessary; as I mentioned in a previous comment, the tester should be able to run the canary on their account/test environment.

File: testing/terraform/ec2/main.tf [hunk 8]
```diff
@@ -27,7 +27,7 @@ locals {
 }
 
 data "aws_ami" "ami" {
-  owners = ["amazon"]
+  owners   = ["amazon"]
```
Comment (majanjua-amzn) [inline testing/terraform/ec2/main.tf:30]: nit: Unnecessary white space change

File: testing/sample-apps/README.md [hunk 9]
```diff
@@ -10,10 +10,16 @@ Ensure that none of the repositories are currently using the image about to be u
...
 
+### Setting up the environment:
+1. Run the `patch.sh` script under `.github/scripts/` folder. You should have a new folder called `opentelemetry-java-instrumentation`
+2. Cd to the new folder, then run `gradle publishToMavenLocal`
+3. Run `rm -rf opentelemetry-java-instrumentation` to delete the folder.
+
 ### Steps to update image:
 1. Use `ada` commands to autheticate into the testing account
-2. Login to ECR Repository: `aws ecr get-login-password --region us-east-1 | docker login --username AWS --password-stdin {REPOSITORY}`
+2. Login to ECR Repository: `aws ecr get-login-password --region us-east-1 | docker login --username AWS --password-stdin {REPOSITORY}`. Create a new repository first if the repository doesn't exist.
```
Comment (majanjua-amzn) [inline testing/sample-apps/README.md:20]: If it's needed, just make that step 1

File: testing/sample-apps/README.md [hunk 10]
```diff
@@ -10,10 +10,16 @@ Ensure that none of the repositories are currently using the image about to be u
...
 3. Change repository name in the `build.gradle.kts` file under `testing/sample-apps/springboot` or `testing/sample-apps/sprintboot-remote-service`
+4. Change the `tasks.named("jib").enabled` value on the `build.gradle.kts` file from false to true
```
Comment (majanjua-amzn) [inline testing/sample-apps/README.md:22]: Why is it set to false in the first place?

File: .github/workflows/appsignals-e2e-eks-test.yml [hunk 11]
```diff
@@ -42,6 +42,24 @@ jobs:
...
+            scripts/eks/appsignals/clean-app-signals.sh
+          sparse-checkout-cone-mode: false
+
+      - name: Remove log group deletion command
+        if: always()
+        working-directory: enablement-script/scripts/eks/appsignals
+        run: |
+          delete_log_group="aws logs delete-log-group --log-group-name '${{ env.LOG_GROUP_NAME }}' --region \$REGION"
+          sed -i "s#$delete_log_group##g" clean-app-signals.sh
+
```
Comment (majanjua-amzn) [inline .github/workflows/appsignals-e2e-eks-test.yml:62]: nit: This should be a separate PR. We're mixing 2 different tasks here and we could've merged this small fix by now.

File: testing/terraform/ec2/main.tf [hunk 12]
```diff
@@ -66,7 +66,7 @@ resource "aws_instance" "main_service_instance" {
   ami                                   = data.aws_ami.ami.id # Amazon Linux 2 (free tier)
   instance_type                         = "t2.micro"
   key_name                              = local.ssh_key_name
-  iam_instance_profile                  = "APP_SIGNALS_EC2_TEST_ROLE"
+  iam_instance_profile                  = var.test_role
```
Comment (majanjua-amzn) [inline testing/terraform/ec2/main.tf:69]: Will this immediately break release testing in CWA and CWA Operator?

File: testing/README.md [hunk 13]
```diff
@@ -0,0 +1,95 @@
...
+  - `kubectl apply -f v2_4_7_full.yaml`
+  - `curl -Lo v2_4_7_ingclass.yaml https://github.com/kubernetes-sigs/aws-load-balancer-controller/releases/download/v2.4.7/v2_4_7_ingclass.yaml`
+  - `kubectl apply -f v2_4_7_ingclass.yaml`
+
+### 3. Setting up Environment for EC2 Tests
+#### Create IAM Role for EC2 Instance
+- Login to AWS, go to the IAM console and click on the Roles tab. Click Create role.
+- Choose AWS service, and choose EC2 as the use case. Click Next.
+- Choose AmazonS3ReadOnlyAccess, AWSXrayWriteOnlyAccess, and CloudWatchAgentServerPolicy as the permission. 
+- Type the role name and click "Create role".
```
Comment (majanjua-amzn) [inline testing/README.md:59]: Let's ask them to name their role the same as we do. This would avoid the need for secret updates everywhere and a change in other repos I believe.

**Review by majanjua-amzn — CHANGES_REQUESTED** [review]
Great work, it's quite detailed. Left some comments, let's make sure we understand the impact on the other repos. For other repos: (1) Do we need to update secrets? Let's do so before merging of course (2) Will this break the release tests? Is a code change required in the other repos as well? Seems like yes based on changes to terraform.

See comment on how we can avoid both.

File: testing/README.md (hunk 3 above)
Comment (harrryr) [reply testing/README.md:52]: Done

File: testing/README.md (hunk 2 above)
Comment (harrryr) [reply testing/README.md:39]: Done

File: testing/README.md (hunk 4 above)
Comment (harrryr) [reply testing/README.md:68]: Done

File: testing/README.md (hunk 5 above)
Comment (harrryr) [reply testing/README.md:72]: Done

File: testing/README.md (hunk 6 above)
Comment (harrryr) [reply testing/README.md:3]: Done

File: testing/README.md (hunk 7 above)
Comment (harrryr) [reply testing/README.md:89]: Is that what we want? Rather than having a canary running on their forked repo isn't it better to just test change when we push new commit? Or do you mean we change the `on:` value to push on the canary file when we want to test?

File: testing/sample-apps/README.md (hunk 10 above)
Comment (harrryr) [reply testing/sample-apps/README.md:22]: Looks like it was changed a while back: https://github.com/aws-observability/aws-otel-java-instrumentation/commit/862695ea62ecc3ee04bc8e6969670151d57019f3

I'm assuming the sample app gradle file was interfering when the ADOT team tries to build their images.

File: .github/workflows/appsignals-e2e-eks-test.yml (hunk 11 above)
Comment (harrryr) [reply .github/workflows/appsignals-e2e-eks-test.yml:62]: Testing this change requires us to merge this PR first so I included it together.

File: testing/README.md (hunk 13 above)
Comment (harrryr) [reply testing/README.md:59]: The CW Agent team has already added the secret for us and we just need to update the variables once this PR is merged. I think this change will make it more flexible and also hide our role name? Lmk

File: testing/terraform/ec2/main.tf (hunk 12 above)
Comment (harrryr) [reply testing/terraform/ec2/main.tf:69]: CWA Operator doesn't use EC2, but it will break CWA

File: testing/README.md (hunk 1 above)
Comment (harrryr) [reply testing/README.md:13]: We can. We also have to enter the console anyways to manually change the edit policy for the role so I think we can just make the programmer do this on the console?

File: testing/README.md (hunk 7 above)
Comment (majanjua-amzn) [reply testing/README.md:97]: It would be best if the tester doesn't have to add files and then remove them after the fact. If we're setting up a test environment it should be made to be exactly like the real environment so that we can just run the same test in our accounts.

File: testing/README.md (hunk 13 above)
Comment (majanjua-amzn) [reply testing/README.md:67]: Showing role name is not a big deal, and I think this will help reduce the work involved in running the test/having to update the other repo.

**Review by majanjua-amzn — APPROVED** [review]
Make sure to run test runs

**Review by bryan-aguilar — APPROVED** [review]
